Go 1.16 "go vet" Issues

<!--
Thanks for submitting a bug report. Please fill the template below,
otherwise we will not be able to process this bug report.
-->

#### Summary
<!-- Summarize the problem in a few sentences: -->

...

#### Steps to Reproduce
<!-- How can we reproduce the problem? -->

```bash
$ go vet ./...
```

<!--
Please upload relevant configuration (as .txt).
If you use the command "ttn-lw-stack config", you can redact sensitive config.
-->

#### What do you see now?
<!-- Please paste terminal output, upload logs (as .txt) or upload screenshots. -->

- [x] `# go.thethings.network/lorawan-stack/v3/pkg/webui` - #3904 

    ```
    pkg/webui/template.go:50:2: struct field tag `name:-` not compatible with reflect.StructTag.Get:bad syntax for struct tag value
    ```

- [x] `go.thethings.network/lorawan-stack/v3/pkg/applicationserver/io/grpc_test`
    ```
    pkg/applicationserver/io/grpc/grpc_test.go:116:6: call to (*T).Fatalf from a non-test goroutine
    pkg/applicationserver/io/grpc/grpc_test.go:119:6: call to (*T).FailNow from a non-test goroutine
    ```

- [x] `go.thethings.network/lorawan-stack/v3/pkg/gatewayserver/io/grpc_test`
    ```
    pkg/gatewayserver/io/grpc/grpc_test.go:132:6: call to (*T).Fatalf from a non-test goroutine
    pkg/gatewayserver/io/grpc/grpc_test.go:135:6: call to (*T).FailNow from a non-test goroutine
    pkg/gatewayserver/io/grpc/grpc_test.go:198:5: call to (*T).Fatalf from a non-test goroutine
    ```

- [x] `go.thethings.network/lorawan-stack/v3/pkg/gatewayserver_test`
    ```
    pkg/gatewayserver/gatewayserver_test.go:791:10: call to (*T).Fatalf from a non-test goroutine
    pkg/gatewayserver/gatewayserver_test.go:880:9: call to (*T).Fatalf from a non-test goroutine
    pkg/gatewayserver/gatewayserver_test.go:936:6: call to (*T).Fatalf from a non-test goroutine
    ```

- [x] `go.thethings.network/lorawan-stack/v3/pkg/errors_test` - #3910
    ```
    pkg/errors/errors_test.go:97:7: the cancel function returned by context.WithDeadline should be called, not discarded, to avoid a context leak
    ```

- [ ] `go.thethings.network/lorawan-stack/v3/pkg/gatewayserver/io/ws_test`
    ```
    pkg/gatewayserver/io/ws/ws_test.go:312:6: call to (*T).Fatalf from a non-test goroutine
    pkg/gatewayserver/io/ws/ws_test.go:370:6: call to (*T).Fatalf from a non-test goroutine
    pkg/gatewayserver/io/ws/ws_test.go:415:7: call to (*T).Fatalf from a non-test goroutine
    pkg/gatewayserver/io/ws/ws_test.go:676:6: call to (*T).Fatalf from a non-test goroutine
    pkg/gatewayserver/io/ws/ws_test.go:1013:7: call to (*T).Fatalf from a non-test goroutine
    pkg/gatewayserver/io/ws/ws_test.go:1330:7: call to (*T).Fatalf from a non-test goroutine
    ```

Note: probably not all of these are related to Go 1.16.

#### What do you want to see instead?
<!-- Please add some examples or mock-ups if applicable. -->

No issues

#### Environment
<!--
Your environment: OS/Browser/Gateway/Device/...? Versions? IDs/EUIs?
Paste the output of "ttn-lw-cli version" or "ttn-lw-stack version" if applicable.
-->

Current v3.11 branch (basically v3.11.0)

Go 1.16

Assigning myself for

```
# go.thethings.network/lorawan-stack/v3/pkg/gatewayserver_test
pkg/gatewayserver/gatewayserver_test.go:791:10: call to (*T).Fatalf from a non-test goroutine
pkg/gatewayserver/gatewayserver_test.go:880:9: call to (*T).Fatalf from a non-test goroutine
pkg/gatewayserver/gatewayserver_test.go:936:6: call to (*T).Fatalf from a non-test goroutine|
```

--- a/pkg/applicationserver/io/grpc/grpc_test.go
+++ b/pkg/applicationserver/io/grpc/grpc_test.go
@@ -113,18 +113,19 @@ func TestAuthentication(t *testing.T) {
 
 			wg := sync.WaitGroup{}
 			wg.Add(1)
+			var err error
 			go func() {
 				defer wg.Done()
-				_, err := client.Subscribe(ctx, &tc.ID, creds)
-				if tc.OK && err != nil && !a.So(errors.IsCanceled(err), should.BeTrue) {
-					t.Fatalf("Unexpected link error: %v", err)
-				}
-				if !tc.OK && !a.So(errors.IsCanceled(err), should.BeFalse) {
-					t.FailNow()
-				}
+				_, err = client.Subscribe(ctx, &tc.ID, creds)
 			}()
-
 			wg.Wait()
+
+			if tc.OK && err != nil && !a.So(errors.IsCanceled(err), should.BeTrue) {
+				t.Fatalf("Unexpected link error: %v", err)
+			}
+			if !tc.OK && !a.So(errors.IsCanceled(err), should.BeFalse) {
+				t.FailNow()
+			}
 		})
 	}
 }

--- a/pkg/gatewayserver/io/ws/ws_test.go
+++ b/pkg/gatewayserver/io/ws/ws_test.go
@@ -21,6 +21,7 @@ import (
 	"math"
 	"net"
 	"net/http"
+	"sync"
 	"testing"
 	"time"
 
@@ -301,15 +302,20 @@ func TestDiscover(t *testing.T) {
 				t.Fatalf("Failed to write message: %v", err)
 			}
 
+			var readErr error
 			resCh := make(chan []byte)
+			var wg sync.WaitGroup
+			wg.Add(1)
 			go func() {
+				defer wg.Done()
 				_, data, err := conn.ReadMessage()
 				if err != nil {
 					close(resCh)
 					if err == websocket.ErrBadHandshake {
 						return
 					}
-					t.Fatalf("Failed to read message: %v", err)
+					readErr = err
+					return
 				}
 				resCh <- data
 			}()
@@ -322,6 +328,10 @@ func TestDiscover(t *testing.T) {
 				a.So(response, should.Resemble, tc.Response)
 			case <-time.After(timeout):
 				t.Fatal("Read message timeout")
+			}
+			wg.Wait()
+			if readErr != nil {
+				t.Fatalf("Failed to read message: %v", readErr)
 			}
 		})
 	}
@@ -364,12 +374,16 @@ func TestDiscover(t *testing.T) {
 				t.Fatalf("Failed to write message: %v", err)
 			}
 
+			var wg sync.WaitGroup
+			wg.Add(1)
 			go func() {
-				_, _, err := conn.ReadMessage()
-				if err == nil {
-					t.Fatalf("Expected connection closure with error but received none")
-				}
+				defer wg.Done()
+				_, _, err = conn.ReadMessage()
 			}()
+			wg.Wait()
+			if err != nil {
+				t.Fatalf("Failed to read message: %v", err)
+			}
 		})
 	}
 
@@ -404,16 +418,20 @@ func TestDiscover(t *testing.T) {
 			if err := conn.WriteMessage(websocket.TextMessage, req); err != nil {
 				t.Fatalf("Failed to write message: %v", err)
 			}
+			var readErr error
 			resCh := make(chan []byte)
+			var wg sync.WaitGroup
+			wg.Add(1)
 			go func() {
+				defer wg.Done()
 				_, data, err := conn.ReadMessage()
 				if err != nil {
 					close(resCh)
 					if err == websocket.ErrBadHandshake {
 						return
-					} else {
-						t.Fatalf("Failed to read message: %v", err)
 					}
+					readErr = err
+					return
 				}
 				resCh <- data
 			}()
@@ -433,7 +451,10 @@ func TestDiscover(t *testing.T) {
 			case <-time.After(timeout):
 				t.Fatalf("Read message timeout")
 			}
-			conn.Close()
+			wg.Wait()
+			if readErr != nil {
+				t.Fatalf("Failed to read message: %v", readErr)
+			}
 		})
 	}
 }
@@ -669,11 +690,16 @@ func TestVersion(t *testing.T) {
 				t.Fatalf("Failed to write message: %v", err)
 			}
 
+			var readErr error
+			var wg sync.WaitGroup
 			resCh := make(chan []byte)
+			wg.Add(1)
 			go func() {
+				defer wg.Done()
 				_, data, err := conn.ReadMessage()
 				if err != nil {
-					t.Fatalf("Failed to read message: %v", err)
+					readErr = err
+					return
 				}
 				resCh <- data
 			}()
@@ -687,6 +713,10 @@ func TestVersion(t *testing.T) {
 				a.So(response, should.Resemble, tc.ExpectedRouterConfig)
 			case <-time.After(timeout):
 				t.Fatalf("Read message timeout")
+			}
+			wg.Wait()
+			if readErr != nil {
+				t.Fatalf("Failed to read message: %v", err)
 			}
 			select {
 			case stat := <-gsConn.Status():
@@ -1039,11 +1069,16 @@ func TestTraffic(t *testing.T) {
 					t.Fatalf("Failed to send downlink: %v", err)
 				}
 
-				resCh := make(chan []byte)
+				var readErr error
+				var wg sync.WaitGroup
+				resCh := make(chan []byte, 1)
+				wg.Add(1)
 				go func() {
+					defer wg.Done()
 					_, data, err := wsConn.ReadMessage()
 					if err != nil {
-						t.Fatalf("Failed to read message: %v", err)
+						readErr = err
+						return
 					}
 					resCh <- data
 				}()
@@ -1062,6 +1097,10 @@ func TestTraffic(t *testing.T) {
 					}
 				case <-time.After(timeout):
 					t.Fatalf("Read message timeout")
+				}
+				wg.Wait()
+				if readErr != nil {
+					t.Fatalf("Failed to read message: %v", readErr)
 				}
 			}
 		})
@@ -1356,11 +1395,16 @@ func TestRTT(t *testing.T) {
 					t.Fatalf("Failed to send downlink: %v", err)
 				}
 
+				var readErr error
+				var wg sync.WaitGroup
 				resCh := make(chan []byte)
+				wg.Add(1)
 				go func() {
+					defer wg.Done()
 					_, data, err := wsConn.ReadMessage()
 					if err != nil {
-						t.Fatalf("Failed to read message: %v", err)
+						readErr = err
+						return
 					}
 					resCh <- data
 				}()
@@ -1377,6 +1421,10 @@ func TestRTT(t *testing.T) {
 					testTime.Rx = &now
 				case <-time.After(timeout):
 					t.Fatalf("Read message timeout")
+				}
+				wg.Wait()
+				if readErr != nil {
+					t.Fatalf("Failed to read message: %v", err)
 				}
 			}
 		})

